### PR TITLE
feat(auth): set up App Store modules

### LINKED ISSUE #10313: Create App Store payment class
Create service class for Apple App Store receipt verification and storage. This class should utilize auth firestore config to prepare for its receipt storage system. It will need configuration information such as the appStoreReceiptURL to use for validating receipts.

This class also needs a method that returns whether the receipt should be considered active or not based on its expiration and trial status.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FXA-3934)
┆Attachments: <a href="https://mozilla-hub.atlassian.net/rest/api/2/attachment/content/22286">2022-03-30_slack-bdanforth.txt</a>

### LINKED ISSUE #12579: [subplat][AppleIAP] Update exports in apple-app-store/api-client
We are using an [App Store Server API client library](https://github.com/agisboye/app-store-server-api) in [https://mozilla-hub.atlassian.net/browse/FXA-3934](https://mozilla-hub.atlassian.net/browse/FXA-3934|smart-link), but the library does not export a number of types we make use of in its `/dist/cjs/index.js` file (e.g. `JWSTransactionDecodedPayload`).

Temporarily, I have copied over all the types from the library into `apple-app-store/api-client`, but we should open a PR in the library's repo to try and get all of them included in the library’s default exports and update our code if the PR is accepted, and the library cuts a new release.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FXA-4920)

### PR DESCRIPTION
**Pull request by biancadanforth**
Because:

* We want SubPlat to know about Apple IAP subscriptions for our RPs.

This commit:

* Sets up the needed Apple IAP modules in lib/payments/iap/apple-app-store including classes, their methods and types, largely modeled after the Google Play IAP modules.
* Adds App Store config (under config.subscriptions.appStore) including a sandbox flag and App Store credentials by app bundleId.
* Moves lib/payments/google-play to lib/payments/iap/google-play.
* Moves IAPConfig class to the new shared lib/payments/iap dir.
* Adds two new libraries to the auth server: app-store-server-api (App Store Server API client) and type-fest (for helpful TS types, such as RequireAtLeastOne used here).

Closes #10313, closes #12579

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [x] I have added necessary documentation (if appropriate).
- [x] I have verified that my changes render correctly in RTL (if appropriate).

## Screenshots (Optional)

Please attach the screenshots of the changes made in case of change in user interface.

## Other information (Optional)

Here are the analogs I modeled the different modules on from the Google Play Store integration:
* AppleIAP ←→ PlayBilling
* SubscriptionPurchase ←→ SubscriptionPurchase
* PurchaseManager ←→ PurchaseManager/UserManager

### CODE REVIEW

**Comment by biancadanforth** [comment]
@bbangert I'm particularly interested in feedback on any parts that I can descope (e.g. if they are or can be covered by separate tickets) in order to land this sooner rather than later and unblock the other tickets.

**Review by biancadanforth — COMMENTED** [review]
@bbangert : Here is the finished patch without tests and some questions for you. It'd be great if you can review this while I am working on writing the tests. I'll ping you in Slack about this as well as we discussed.

Thank you in advance; I know this is a very large patch.

File: packages/fxa-auth-server/config/index.ts
Comment (biancadanforth) [inline packages/fxa-auth-server/config/index.ts:788]: Unfortunately, `node-convict` doesn't support keys with dots in them (https://github.com/mozilla/node-convict/issues/250), so `yarn start` fails.

I've replaced the dots here with underscores, but that means we'd have make a copy of the `bundleId`s with this string replace operation to get API clients keyed off them. Not a great solution. Would you have any suggestions?

File: packages/fxa-auth-server/lib/payments/apple-app-store/apple-iap.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/apple-app-store/apple-iap.ts]: I did what I proposed in [2ac3224](https://github.com/mozilla/fxa/pull/12374/commits/2ac3224b9bf38ac7394767467b3e97329793a923).

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/types/api-client.ts
Comment (bbangert) [inline packages/fxa-auth-server/lib/payments/iap/apple-app-store/types/api-client.ts:28]: That's odd, I'd be curious if we can fix this easily.

File: packages/fxa-auth-server/lib/payments/apple-app-store/subscription-purchase.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/apple-app-store/subscription-purchase.ts]: Talked with Ben: Since grace period is always less than billing retry, we can ignore that check here.

File: packages/fxa-auth-server/lib/payments/apple-app-store/subscription-purchase.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/apple-app-store/subscription-purchase.ts]: I'll take another pass through the slides on the Apple videos, otherwise we may have to check later with a test subscription that exhausts billing retry.

File: packages/fxa-auth-server/lib/payments/apple-app-store/apple-iap.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/apple-app-store/apple-iap.ts]: This sounds reasonable per Ben provided they have sufficient shared things.

File: packages/fxa-auth-server/config/index.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/config/index.ts]: Ben suggested using the new `getBundleId` method and key off `appName` instead, so I'll do that.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/types/api-client.ts
Comment (biancadanforth) [inline packages/fxa-auth-server/lib/payments/iap/apple-app-store/types/api-client.ts:27]: @bbangert : I couldn't get rid of this file after all; see code comments.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/apple-iap.ts
Comment (biancadanforth) [inline packages/fxa-auth-server/lib/payments/iap/apple-app-store/apple-iap.ts:42]: This convict workaround isn't pretty.

The challenge is that `getBundleId` is async, while the App Store Server API client is initialized in the `constructor` of the `AppleIAP` class.

Barry had a similar challenge in the `PaymentConfigurationManager` where he [calls the async `load` method in the `constructor`](https://github.com/mozilla/fxa/blob/0f9936330fb31a2cb0cbafe675a46c4ee632598b/packages/fxa-auth-server/lib/payments/configuration/manager.ts#L42), but in his case, his `maybeLoad` helper was only used internally by the `PaymentConfigManager` in known/finite ways. Here, the API client is used by an external module, `PurchaseManager`, and we’d have to remember to add this check for any future method we may add that makes use of the API client.

In a similar vein, I ended up creating a new static, async method on the `AppleIAP` class called `initializeApiClientByBundleId`, and I added a check in the client's caller, `PurchaseManager.querySubscriptionPurchase` to call this method if the client isn’t yet initialized.

Is there a simpler way to do this? Maybe it's okay to s/_/.? That logic was a lot more readable/straightforward.

**Review by biancadanforth — COMMENTED** [review]
Almost done with tests. I fixed (or attempted to in a couple of cases; see inline comments) the issues we discussed earlier today.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/apple-iap.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/iap/apple-app-store/apple-iap.ts:42]: Seems like an antipattern, but I do have yet another idea/option: I thought perhaps if I kept the config keyed off `bundleId`, I could swap out the dots just before `conf.validate()` is called and then swap them back in afterward as a hack.

@bbangert , do you have a preference for any of these three approaches (initialize API client as needed before using, use underscores and string.replace just before creating the API client, or string.replace before/after `conf.validate()`) or any other suggestions?

**Comment by biancadanforth** [comment]
I think we disabled squash and merge, but I can do an interactive rebase and roll these commits into one commit before merging.

I will file a follow-up ticket to set the App Store credentials/sandbox env vars in stage and prod and open a cloudops PR.

File: packages/fxa-auth-server/lib/payments/apple-app-store/subscription-purchase.ts
Comment (bbangert) [reply packages/fxa-auth-server/lib/payments/apple-app-store/subscription-purchase.ts]: A new subscription purchase would be needed if the original transaction id was different. That's the primary key of these, so if a purchase shares that, than its linked to the original so it updates the existing one, etc.

**Review by bbangert — DISMISSED** [review]

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/apple-iap.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/iap/apple-app-store/apple-iap.ts:52]: A static method means an instance of the class is not required; it's related to the class in the sense that the class' constructor is what would ordinarily initialize the API client.

If we're going with the string.replace underscore approach though, this method will be removed anyway.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts
Comment (bbangert) [inline packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts:67]: This is mixing up concerns, as the class that wants to make an API call is now responsible for handling the API client pool. The Apple IAP client should have a helper service class that is responsible for using the appropriate `bundleId`, creating one if needed, then using it. So that the API instead is just:
```typescript
apiResponse = await this.appStoreClient.getSubscriptionStatuses(bundleId, originalTransactionId);
```

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/apple-iap.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/iap/apple-app-store/apple-iap.ts:39]: This was removed since I went with the underscore and string.replace approach.

If it were still in place, if this failed, the API client would not be initialized, and a check in the caller would have called the `initializeApiClientByBundleId` helper.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/apple-iap.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/iap/apple-app-store/apple-iap.ts:27]: Yeah I saw that and was planning to do a follow-up to refactor, but I can roll that one into this PR instead.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/errors.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/iap/apple-app-store/errors.ts:13]: I modeled this off of the [`PaypalClientError` class](https://github.com/mozilla/fxa/blob/39a94ee69c406981a9f003621205a4541eb5a59b/packages/fxa-auth-server/lib/payments/paypal/client.ts#L257-L267), but I can model it more like the [`app-store-server-api` `AppStoreError` class](https://github.com/agisboye/app-store-server-api/blob/33c466733a223974e077264a525ef36c4324f5de/src/Errors.ts#L1-L8).

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/subscription-purchase.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/iap/apple-app-store/subscription-purchase.ts:214]: I am basing this off of the Apple docs for [`offerType`](https://developer.apple.com/documentation/appstoreserverapi/offertype) and what they mean by an introductory offer (linked under the "Discussion" section).

The word "free trial" only shows up in the docs related to introductory offers. I'm happy to remove this helper for now and add it later on once we know how the VPN team is implementing free trials.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts:67]: I've removed this `if` block (and the method within) in favor of the string.replace approach, but I agree with you that this was not ideal.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/types/api-client.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/iap/apple-app-store/types/api-client.ts]: Ben and I discussed this in Slack; I filed https://mozilla-hub.atlassian.net/browse/FXA-4920.

Per Ben:
> My main worry about copying the types in, is it makes our code brittle. A dependabot push to bump our use of his library can easily break things if he changes the types at all if they're not a public API from the library.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/apple-iap.ts
Comment (bbangert) [inline packages/fxa-auth-server/lib/payments/iap/apple-app-store/apple-iap.ts:42]: What is in charge of managing which bundleId to use for which API request? I thought this was going to be folded into a class that handles it automatically.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts
Comment (bbangert) [inline packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts:69]: Having a single class that ensures the bundleId exists, uses the appropriate one, etc. would make this a bit cleaner and provide more flexibility in the future.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts
Comment (bbangert) [inline packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts:248]: Why does the entitlement not being active mean that it could've changed?

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts
Comment (bbangert) [inline packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts:76]: This is kind of a bummer, I looked at the implementation, its identical for each of these which means a bit of repeat work. Hopefully its not re-fetching certs as well but I'm not entirely sure. Putting a FIXME or TODO above this pointing out possible inefficiency and optimization here would be good.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/types/api-client.ts
Comment (bbangert) [inline packages/fxa-auth-server/lib/payments/iap/apple-app-store/types/api-client.ts:28]: Maybe a TODO here with a reference to the other library that this can be removed when they export it?

Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/iap/apple-app-store/types/api-client.ts:28]: The PR I opened in the app-store-client-api repo was merged and the maintainer tagged a new release, so I plan to update the package in this PR and remove this file (i.e. close #12579 at the same time).

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/apple-iap.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/iap/apple-app-store/apple-iap.ts:42]: As we discussed, I created a new class to handle this, `AppStoreHelper`.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts:69]: I created the `AppStoreHelper` class.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts:248]: We discussed this sync; I updated the comment to be more clear about why we are doing this.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts:76]: I'll file a follow-up to look into this more and link it here.

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/app-store-helper.ts
Comment (bbangert) [inline packages/fxa-auth-server/lib/payments/iap/apple-app-store/app-store-helper.ts:52]: If the bundleId is not in the credentials, this will throw a TypeError and we'll return a 500 error which might not be obvious when debugging on Sentry about why it happened (a mismatch between bundleId defined in Firestore, and the one in auth-server config).

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts
Comment (bbangert) [inline packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts:255]: 👍🏻

File: packages/fxa-auth-server/lib/payments/iap/apple-app-store/app-store-helper.ts
Comment (biancadanforth) [reply packages/fxa-auth-server/lib/payments/iap/apple-app-store/app-store-helper.ts:52]: Thanks for pointing this out -- I've done the following -- is this what you had in mind for a more descriptive error?

```diff
bdanforth ~/Projects/fxa/packages/fxa-auth-server (fxa-3934-apple-iap-base-class) $ cat ~/fxa-3934-wip.diff 
diff --git a/packages/fxa-auth-server/lib/payments/iap/apple-app-store/app-store-helper.ts b/packages/fxa-auth-server/lib/payments/iap/apple-app-store/app-store-helper.ts
index 9d8ccbc1c1..ed5789f749 100644
--- a/packages/fxa-auth-server/lib/payments/iap/apple-app-store/app-store-helper.ts
+++ b/packages/fxa-auth-server/lib/payments/iap/apple-app-store/app-store-helper.ts
@@ -9,6 +9,7 @@ import {
 import { Container } from 'typedi';
 
 import { AppConfig, AuthLogger } from '../../../types';
+import { AppStoreHelperError } from './types/errors';
 
 export class AppStoreHelper {
   private log: AuthLogger;
@@ -48,6 +49,13 @@ export class AppStoreHelper {
     if (this.appStoreServerApiClients.hasOwnProperty(bundleId)) {
       return this.appStoreServerApiClients[bundleId];
     }
+    if (!this.credentialsByBundleId.hasOwnProperty(bundleId)) {
+      const libraryError = new Error(
+        `No App Store credentials found for app with bundleId: ${bundleId}.`
+      );
+      libraryError.name = AppStoreHelperError.CREDENTIALS_NOT_FOUND;
+      throw libraryError;
+    }
     const { serverApiKey, serverApiKeyId, issuerId } =
       this.credentialsByBundleId[bundleId];
     this.appStoreServerApiClients[bundleId] = new AppStoreServerAPI(
diff --git a/packages/fxa-auth-server/lib/payments/iap/apple-app-store/types/errors.ts b/packages/fxa-auth-server/lib/payments/iap/apple-app-store/types/errors.ts
index f036e122ec..74f905964d 100644
--- a/packages/fxa-auth-server/lib/payments/iap/apple-app-store/types/errors.ts
+++ b/packages/fxa-auth-server/lib/payments/iap/apple-app-store/types/errors.ts
@@ -14,3 +14,7 @@ export enum PurchaseUpdateError {
   INVALID_ORIGINAL_TRANSACTION_ID = 'InvalidOriginalTransactionId',
   OTHER_ERROR = 'OtherError',
 }
+
+export enum AppStoreHelperError {
+  CREDENTIALS_NOT_FOUND = 'CredentialsNotFound',
+}
```

**Pull request merged**
